pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -456,3 +484,317 @@ func (plugin *kubenetNetworkPlugin) getNsenterPath() (string, error) {
+		if err != nil {
+			glog.V(3).Info("Failed to retrieve pod ip for %s-%s: %v", p.Namespace, p.Name, err)
+			continue
+		}
+		// Need the complete api.Pod object
+		pod, ok := plugin.host.GetPodByName(p.Namespace, p.Name)
+		if ok {
+			for _, container := range pod.Spec.Containers {
+				for _, port := range container.Ports {
+					if port.HostPort != 0 {
+						podHostportMap[port] = targetPod{podFullName: kubecontainer.GetPodFullName(pod), podIP: podIP.String()}
+					}
+				}
+			}
+		}
+	}
+	return podHostportMap, nil
+}
+
+// Join all words with spaces, terminate with newline and write to buf.
+func writeLine(buf *bytes.Buffer, words ...string) {
+	buf.WriteString(strings.Join(words, " ") + "\n")
+}
+
+//hostportChainName takes containerPort for a pod and returns associated iptables chain.
+// This is computed by hashing (sha256)
+// then encoding to base32 and truncating with the prefix "KUBE-SVC-".  We do
+// this because Iptables Chain Names must be <= 28 chars long, and the longer
+// they are the harder they are to read.
+func hostportChainName(cp api.ContainerPort, podFullName string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(string(cp.HostPort) + string(cp.Protocol) + podFullName))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain(kubenetHostportChainPrefix + encoded[:16])
+}
+
+// syncHostportsRules gathers all hostports on node and setup iptables rules enable them. And finally clean up stale hostports
+func (plugin *kubenetNetworkPlugin) syncHostportsRules() {
+	start := time.Now()
+	defer func() {
+		glog.V(4).Infof("syncHostportsRules took %v", time.Since(start))
+	}()
+
+	containerPortMap, err := plugin.gatherAllHostports()
+	if err != nil {
+		glog.Errorf("Fail to get hostports: %v", err)
+		return
+	}
+
+	glog.V(4).Info("Ensuring kubenet hostport chains")
+	// Ensure kubenetHostportChain
+	if _, err := plugin.iptables.EnsureChain(utiliptables.TableNAT, kubenetHostportChain); err != nil {
+		glog.Errorf("Failed to ensure that %s chain %s exists: %v", utiliptables.TableNAT, kubenetHostportChain, err)
+		return
+	}
+	tableChainsNeedJumpServices := []struct {
+		table utiliptables.Table
+		chain utiliptables.Chain
+	}{
+		{utiliptables.TableNAT, utiliptables.ChainOutput},
+		{utiliptables.TableNAT, utiliptables.ChainPrerouting},
+	}
+	args := []string{"-m", "comment", "--comment", "kubenet hostport portals",
+		"-m", "addrtype", "--dst-type", "LOCAL",
+		"-j", string(kubenetHostportChain)}
+	for _, tc := range tableChainsNeedJumpServices {
+		if _, err := plugin.iptables.EnsureRule(utiliptables.Prepend, tc.table, tc.chain, args...); err != nil {
+			glog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", tc.table, tc.chain, kubenetHostportChain, err)
+			return
+		}
+	}
+	// Need to SNAT traffic from localhost
+	args = []string{"-m", "comment", "--comment", "kubenet postrouting rules", "-o", BridgeName, "-s", "127.0.0.1/8", "-j", "MASQUERADE"}
```

Comment:
did you try accessing a host-port from the container that services that host-port?  It might need SNAT, too